src/ir_def/construct.ml

```diff
@@ -600,22 +607,48 @@ let loopWhileE exp1 exp2 =
         )
     )
 
+let countingE binding arr expSize expIndexing expWorker =
+  (* when e1 = arr.vals()
+
+     for p in e1 e2
+     ~~>
+     let arr = ... ;
+     var indx = 0 ;
+     let size = arr.size() ;
+     label l loop {
+       if indx < size
+       then { let p = arr[indx]; e2; indx += 1 }
+       else { break l }
+     } *)
+  let indx = fresh_var "indx" T.(Mut nat) in
```

Comment:
Maybe assert that `arr` has the expected type, as safeguard and documentation?

Reply:
It gets implicitly asserted in `rewrite_for_to_while`. This function is gone now.